tools/witx/src/docs/ast.rs

```diff
@@ -0,0 +1,487 @@
+            node.content_ref_mut::<MdFunc>()
+                .inputs
+                .push((param.name.as_str().to_owned(), param.tref.type_name()));
+        }
+
+        node.new_child(MdSection::new(header.as_str(), "Results"));
+        for result in &self.results {
+            let name = result.name.as_str();
+            let id = if let Some(id) = node.any_ref().id() {
+                format!("{}.{}", id, name)
+            } else {
+                name.to_owned()
+            };
+            let child = node.new_child(MdNamedType::new(
+                "-",
+                id.as_str(),
+                name,
+                result.name.as_str(),
+            ));
+            result.generate(child.clone());
+            // TODO should this be expanded recursively instead of using flattened type names?
+            node.content_ref_mut::<MdFunc>()
+                .outputs
+                .push(result.tref.type_name());
+        }
+    }
+}
+
+impl ToMarkdown for InterfaceFuncParam {
+    fn generate(&self, node: MdNodeRef) {
+        self.tref.generate(node.clone());
+        node.content_ref_mut::<MdNamedType>().docs = self.docs.clone();
+    }
+}
+
+impl BuiltinType {
+    pub fn type_name(&self) -> &'static str {
+        match self {
+            BuiltinType::String => "string",
+            BuiltinType::Char8 => "char8",
+            BuiltinType::USize => "usize",
+            BuiltinType::U8 => "u8",
+            BuiltinType::U16 => "u16",
+            BuiltinType::U32 => "u32",
+            BuiltinType::U64 => "u64",
+            BuiltinType::S8 => "s8",
+            BuiltinType::S16 => "s16",
+            BuiltinType::S32 => "s32",
+            BuiltinType::S64 => "s64",
+            BuiltinType::F32 => "f32",
+            BuiltinType::F64 => "f64",
+        }
+    }
+}
+
+impl TypeRef {
+    pub fn type_name(&self) -> String {
+        match self {
+            TypeRef::Name(n) => n.name.as_str().to_string(),
+            TypeRef::Value(ref v) => match &**v {
+                Type::Array(a) => format!("Array<{}>", a.type_name()),
+                Type::Pointer(p) => format!("Pointer<{}>", p.type_name()),
+                Type::ConstPointer(p) => format!("ConstPointer<{}>", p.type_name()),
+                Type::Builtin(b) => b.type_name().to_string(),
+                Type::Enum { .. }
+                | Type::Int { .. }
+                | Type::Flags { .. }
+                | Type::Struct { .. }
+                | Type::Union { .. }
+                | Type::Handle { .. } => {
+                    unimplemented!("type_name of anonymous compound datatypes")
+                }
+            },
+        }
+    }
+}
+
+impl IntRepr {
+    fn type_name(&self) -> &'static str {
+        match self {
+            IntRepr::U8 => "u8",
+            IntRepr::U16 => "u16",
+            IntRepr::U32 => "u32",
+            IntRepr::U64 => "u64",
+        }
+    }
+}
+
+// TODO
```

Comment:
this TODO is done right?

Reply:
No, not yet. I purposely didn't touch the polyfill docs just yet, as I wanted to focus on the main docs first, and do the polyfill in a subsequent PR. Does that sound OK?

Reply:
My bad I misread what the todo meant. I see now. Its fine to wait on the polyfill docs, they're not very important.